dev/PushNotifications/PushNotificationsLongRunningTask.ProxyStub/NotificationsLongRunningProcess.idl

```diff
@@ -11,7 +11,7 @@ import "ocidl.idl";
 [pointer_default(unique)]
 interface IWpnForegroundSink : IUnknown
 {
-    HRESULT InvokeAll([in] ULONG length, [in, size_is(length)] byte* data, [out] BOOL* foregroundHandled);
+    HRESULT InvokeAll([in] ULONG length, [in, size_is(length)] byte* data, [in] LPCWSTR correlationVector, [out] BOOL* foregroundHandled);
```

Comment:
Is there a chance that the caller of InvokeAll hits the older version of this API?

Reply:
This is also a concern of mine. I'd like to see if @DrusTheAxe has an opinion on this

Reply:
You're changing `InvokeAll`'s footprint? Why? Seems like a Breaking Change.

When was InvokeAll created? 1.1.0?

Who's directly using the proxy? Who's directly using the stub?

If one side's in the Framework package and the other's in the Singleton then yes, you could have client using 1.1.0 framework and server in Singleton using 1.2.0 -- they expect to communicate but don't agree on the footprint. BadMojo(TM), no?

You could ADD a new `IWpnForegroundSink2` with the new footprint (though it's COM so no method overloading - `InvokeAll2()` or `InvokeAllWithCorrelationVector()` or...), support both (old+new) in the server and change the new client to use the new interface.

Reply:
That's what I was affraid of. I had some hope that because it's not a public API and that the components are both part of / delivered with the SDK that maybe this wouldn't be breacking... but yeah, I'll add a new API and support both in the server.

thanks

Reply:
> I had some hope that because it's not a public API and that the components are both part of / delivered with the SDK

If they were both in the same package AND not exposed outward then yes. Client+Server would always be matched ('because you can't take half of package-v1 and half of packge-v2).

But PackagedCOM registers in the COM catalog so folks outside the package can use it. That means client+server aren't guaranteed to be in the same atomic distribution unit (package). And if all uses were you'd almost certainly use WinRT instead of PackagedCOM OOP Server -- if no need to share the server to clients outside the package then an OOP WinRT server would be equivalent but simpler.

>I'll add a new API and support both in the server.

Only caveat is if fwk v2 is installed but singleton is v1, your client call will get an error (probably E_NOINTERFACE). You may want to put an explicit check and logging and/or insights into your call if this happens to guide devs/users to recognize the issue and take corrective action.

For instance, if process is running with Singleton v1's identity and someone tries to AddPackage Singleton-v2 _without_ the 'force' option, we see the Singleton's in use so servicing would be disruptive and fail. Then an app built with 1.2 uses 1.2 fwk but tries to talk to 1.1 singleton and error.

For such logging I recommend a combination of WIL `LOG_IF_FAILED[_MSG]`, `TraceLoggingWrite` and write an event to the event log. See `MddBootstrapInitialize_Log()` in dev\WindowsAppRuntime_BootstrapDLL\MddBootstrap.cpp for an example.

Reply:
P.S. I'd considered making a function in `dev\common` for the common parts but RegisterEventSource+ReportEvent+DeregisterEventSource is so simple but highly variable parameters I didn't see much value. I guess you could create `dev\common\WindowsAppRuntime.EventLog.h`

```c++
namespace wil
{
    typedef unique_any_handle_null<decltype(&::DeregisterEventSource), ::DeregisterEventSource> unique_event_source;
}

namespace Microsoft::WindowsAppRuntime::EventLog
{
inline HRESULT Log(
    const HRESULT eventId,
    _In_ PCWSTR message,
    const WORD severity = EVENTLOG_ERROR_TYPE)
{
    wil::unique_event_source eventlog{ RegisterEventSourceW(nullptr, L"Windows App Runtime");
    RETURN_LAST_ERROR_NUL_NULL(eventlog);

    PCWSTR strings[2]{ L"Windows App Runtime", message };
    return LOG_IF_WIN32_BOOL_FALSE(ReportEventW(eventlog.get(), severity, 0, eventId, nullptr, ARRAYSIZE(strings), 0, strings, nullptr));
}
```

then you can simply call

```
const auto hr{ LOG_IF_FAILED(JustDoIt(x, y)) };
if (FAILED(hr))
{
    auto message{ wil::str_printf<unique_process_heap>(L"Do %s when %y = BadMojo(TM)", x, y) };
    LOG_IF_FAILED(Microsoft::WindowsAppRuntime::EventLog::Log(hr, message.get());
    RETURN_HR(hr);
}
```

Reply:
ok, so that's no fun. Making new version of the IWpnForegroundSink that has an InvokeAllWithCorrelationVector and supporting both implementation is easy enough but it doesn't stop there. The client registers this API has a sync with the server so the server can notify the client when a notification is received.

That means creating a new version of INotificationsLongRunningPlatform with a new version of RegisterForegroundActivator that take a IWpnForegroundSink2 (our new API that supports InvokeAllWithZCorrelationVector).

Which in turn means exposing that new INotificationsLongRunningPlatform to the Library.

Reply:
This is getting we too complicated and I don;t want to jeopardize the entire telemetry feature for InvovkeAll so,I have removed the changes to the IDL from this PR, and I will open a separate PR just for that.